File: src/qmemory/workbench.py
"""Loopback-only read workspace. No MemoryService constructors, migrations or sync side effects."""
from __future__ import annotations

import argparse
import hashlib
import hmac
import json
import secrets
import sqlite3
from contextlib import contextmanager
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

from .config import default_home
from .redaction import redact_text
from .sources import CodexConversationArchive


def clean(value):
    return redact_text(str(value or ""))[0]


class ReadWorkspace:
    def __init__(self, home: Path):
        self.home = home.resolve()

    @contextmanager
    def db(self, name="sources.sqlite3"):
        path = self.home / "state" / name
        if not path.is_file():
            raise FileNotFoundError("index_unavailable")
        connection = sqlite3.connect(path.as_uri() + "?mode=ro", uri=True, timeout=3)
        connection.row_factory = sqlite3.Row
        try:
            connection.execute("PRAGMA query_only=ON")
            connection.execute("BEGIN")
            yield connection
        finally:
            connection.close()

    def projects(self):
        with self.db() as db:
            rows = db.execute("""SELECT p.project_id,p.display_name,p.canonical_root,p.remote,COUNT(c.conversation_id) AS source_count,
                MAX(c.imported_at) AS indexed_at FROM projects p LEFT JOIN conversations c
                ON c.project_id=p.project_id WHERE p.ignored=0 GROUP BY p.project_id
                ORDER BY indexed_at DESC""").fetchall()
        return {"projects": [{"id": r["project_id"], "name": clean(r["display_name"] or
                 (Path(r["canonical_root"]).name if r["canonical_root"] else "") or
                 (str(r["remote"]).rstrip("/").rsplit("/", 1)[-1].removesuffix(".git") if r["remote"] else "") or "未命名项目"),
                 "identity_kind": "repository" if r["project_id"].startswith("git:") else "directory",
                 "source_count": r["source_count"], "indexed_at": r["indexed_at"]} for r in rows],
                "mode": "local_readonly", "cloud_connected": False}

    @staticmethod
    def groups(db, project):
        if not db.execute("SELECT 1 FROM projects WHERE project_id=? AND ignored=0", (project,)).fetchone():
            raise LookupError("project_not_found")
        rows = {r["conversation_id"]: dict(r) for r in db.execute(
            "SELECT conversation_id,source_kind,title,updated_at,imported_at,duplicate_of,parent_conversation_id,mirror_sha256,message_count "
            "FROM conversations WHERE project_id=?", (project,))}
        groups = {}
        for cid, row in rows.items():
            root = CodexConversationArchive._task_root_id(cid, rows)
            groups.setdefault(root, []).append(row)
        return groups

    @staticmethod
    def summary(task_id, members):
        root = next((r for r in members if r["conversation_id"] == task_id), members[0])
        return {"id": task_id, "title": clean(root["title"]), "source_count": len(members),
                "agents": sorted({r["source_kind"] for r in members}),
                "branch_count": sum(bool(r["parent_conversation_id"]) for r in members),
                "updated_at": max((r["updated_at"] or "" for r in members), default=""),
                "work_status": "unverified"}

    def tasks(self, project):
        with self.db() as db:
            groups = self.groups(db, project)
        values = [self.summary(k, v) for k, v in groups.items()]
        values.sort(key=lambda t: t["updated_at"], reverse=True)
        return {"tasks": values, "total": len(values), "relationship_basis": "existing_archive_links"}

    def detail(self, project, task):
        with self.db() as db:
            members = self.groups(db, project).get(task)
            if members is None:
                raise LookupError("task_not_found")
            sources = []
            for row in members:
                cid = row["conversation_id"]
                count = db.execute("SELECT COUNT(*) FROM conversation_messages WHERE conversation_id=?", (cid,)).fetchone()[0]
                sources.append({"id": cid, "agent": row["source_kind"], "version": row["mirror_sha256"],
                    "relation": "branch" if row["parent_conversation_id"] else "copy" if row["duplicate_of"] else "root",
                    "title": clean(row["title"]), "message_count": count, "indexed_at": row["imported_at"]})
        memories, available = [], True
        try:
            with self.db("qmemory.sqlite3") as db:
                refs = {"%s://%s#" % (r["source_kind"], r["conversation_id"].removeprefix("claude:")) for r in members}
                rows = db.execute("SELECT memory_id,statement,subject,status,source_ref,as_of,superseded_by FROM memories WHERE project_id=?", (project,))
                for r in rows:
                    if any(str(r["source_ref"] or "").startswith(ref) for ref in refs):
                        memories.append({k: clean(r[k]) for k in r.keys()})
        except FileNotFoundError:
            available = False
        return {**self.summary(task, members), "sources": sources, "memories": memories,
                "memory_index_available": available, "story_status": "not_generated",
                "scope": {"space_id": "local-private", "project_id": project, "task_id": task},
                "sharing": {"visibility": "private", "published": False}}

    def messages(self, project, task, source, offset=0, sequence=None):
        with self.db() as db:
            members = self.groups(db, project).get(task, [])
            row = next((r for r in members if r["conversation_id"] == source), None)
            if row is None:
                raise LookupError("source_not_in_task")
            total = db.execute("SELECT COUNT(*) FROM conversation_messages WHERE conversation_id=?", (source,)).fetchone()[0]
            if sequence is not None:
                if not db.execute("SELECT 1 FROM conversation_messages WHERE conversation_id=? AND sequence=?", (source, sequence)).fetchone():
                    raise LookupError("evidence_not_found")
                offset = db.execute("SELECT COUNT(*) FROM conversation_messages WHERE conversation_id=? AND sequence<?", (source, sequence)).fetchone()[0]
            rows = db.execute("SELECT sequence,role,text,occurred_at FROM conversation_messages WHERE conversation_id=? ORDER BY sequence LIMIT 30 OFFSET ?", (source, offset)).fetchall()
            items = []
            for m in rows:
                text = clean(m["text"])
                identity = json.dumps([project, source, row["mirror_sha256"], m["sequence"]])
                items.append({"evidence_id": hashlib.sha256(identity.encode()).hexdigest(),
                    "source_ref": "%s://%s#message-%d" % (row["source_kind"], source.removeprefix("claude:"), m["sequence"]),
                    "sequence": m["sequence"], "role": m["role"], "text": text[:20000],
                    "truncated": len(text) > 20000, "occurred_at": m["occurred_at"]})
        return {"messages": items, "offset": offset, "total": total, "next_offset": offset + len(items) if offset + len(items) < total else None,
                "version": row["mirror_sha256"], "view": "redacted_current_index"}


def server(home: Path, port=8790):
    workspace, token = ReadWorkspace(home), secrets.token_urlsafe(32)
    static = Path(__file__).parent / "web"

    class Handler(BaseHTTPRequestHandler):
        def log_message(self, *_):
            pass  # No private project IDs, source refs, credentials or text in access logs.

        def send(self, value, status=200, content_type="application/json"):
            data = json.dumps(value, ensure_ascii=False).encode() if content_type == "application/json" else value
            self.send_response(status)
            self.send_header("Content-Type", content_type + "; charset=utf-8")
            self.send_header("Content-Length", str(len(data)))
            self.send_header("Cache-Control", "no-store")
            self.send_header("X-Content-Type-Options", "nosniff")
            self.send_header("Referrer-Policy", "no-referrer")
            self.send_header("Content-Security-Policy", "default-src 'none'; script-src 'self'; style-src 'self'; connect-src 'self'; frame-ancestors 'none'; base-uri 'none'; form-action 'none'")
            self.end_headers()
            self.wfile.write(data)

        def do_GET(self):
            origin = "http://127.0.0.1:%d" % self.server.server_port
            if self.headers.get("Host") != origin.removeprefix("http://") or self.headers.get("Origin", origin) != origin or self.headers.get("Sec-Fetch-Site") == "cross-site":
                return self.send({"error": "origin_rejected"}, 403)
            url = urlparse(self.path)
            if url.path in ("/", "/qmemory.html"):
                return self.send((static / "workbench.html").read_bytes().replace(b"__SESSION__", token.encode()), content_type="text/html")
            if url.path in ("/workbench.js", "/workbench.css"):
                return self.send((static / url.path[1:]).read_bytes(), content_type="application/javascript" if url.path.endswith(".js") else "text/css")
            if not hmac.compare_digest(self.headers.get("X-QMemory-Session", "").encode("utf-8"), token.encode("utf-8")):
                return self.send({"error": "session_required"}, 401)
            try:
                args = parse_qs(url.query, max_num_fields=8)
                def param(key):
                    values = args.get(key)
                    if not values or len(values) != 1 or len(values[0]) > 1024:
                        raise ValueError("invalid_parameter")
                    return values[0]
                if url.path == "/api/projects":
                    data = workspace.projects()
                elif url.path == "/api/tasks":
                    data = workspace.tasks(param("project"))
                elif url.path == "/api/task":
                    data = workspace.detail(param("project"), param("task"))
                elif url.path == "/api/messages":
                    offset = int(param("offset")) if "offset" in args else 0
                    sequence = int(param("sequence")) if "sequence" in args else None
                    if not 0 <= offset < 2**63 or (sequence is not None and not 1 <= sequence < 2**63):
                        raise ValueError("invalid_parameter")
                    data = workspace.messages(param("project"), param("task"), param("source"), offset, sequence)
                else:
                    return self.send({"error": "not_found"}, 404)
                self.send(data)
            except (OSError, sqlite3.Error):
                self.send({"error": "index_unavailable", "hint": "请先用 QMemory 同步生成索引；本工作台不会迁移或重建。"}, 503)
            except LookupError:
                self.send({"error": "not_found_or_out_of_scope"}, 404)
            except ValueError:
                self.send({"error": "invalid_parameter"}, 400)

        def do_POST(self):
            self.send({"error": "readonly"}, 405)
        do_PUT = do_PATCH = do_DELETE = do_POST

    return ThreadingHTTPServer(("127.0.0.1", port), Handler)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--home", type=Path, default=default_home())
    parser.add_argument("--port", type=int, default=8790)
    args = parser.parse_args()
    app = server(args.home, args.port)
    print("QMemory local read-only workspace: http://127.0.0.1:%d/qmemory.html" % app.server_port, flush=True)
    try:
        app.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        app.server_close()


if __name__ == "__main__":
    main()

File: src/qmemory/web/workbench.js
'use strict';
const $=s=>document.querySelector(s),esc=s=>String(s??'').replace(/[&<>"']/g,c=>({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[c]));
const token=$('meta[name=qmemory-session]').content;
const state={project:new URLSearchParams(location.search).get('project')||'',task:'',view:'board',tasks:[],detail:null,offset:0,next:null,source:'',seq:null,request:0,messageRequest:0};
const agent=s=>s==='claude-code'?'Claude Code':s==='codex'?'Codex':s;
const relation=s=>({branch:'执行分支',copy:'关联副本',root:'主来源'}[s]||s);
const status=s=>({active:'已确认 · 有效',proposed:'AI 候选 · 待核对',superseded:'已被替代',archived:'已归档'}[s]||s);
async function api(path,args={}){const r=await fetch('/api/'+path+'?'+new URLSearchParams(args),{headers:{'X-QMemory-Session':token},cache:'no-store'});const data=await r.json();if(!r.ok)throw Error(data.hint||data.error);return data}
function fail(e){$('#notice').textContent='未能读取：'+e.message+'。原始记录未修改。'}
function taskRows(){const q=$('#search').value.toLocaleLowerCase();return state.tasks.filter(t=>t.title.toLocaleLowerCase().includes(q))}
function title(t){return `<strong>${esc(t.title)}</strong><small>${t.agents.map(agent).map(esc).join(' · ')} · ${t.source_count} 份来源${t.branch_count?' · '+t.branch_count+' 个分支':''}</small>`}
function render(){
 $('#tasks').innerHTML=taskRows().map(t=>`<button data-task="${esc(t.id)}" class="${t.id===state.task?'active':''}">${title(t)}</button>`).join('');
 document.querySelectorAll('[data-view]').forEach(b=>{b.classList.toggle('active',b.dataset.view===state.view);b.setAttribute('aria-pressed',b.dataset.view===state.view?'true':'false')});
 $('#count').textContent=state.tasks.length+' 项任务';
 const d=state.detail;
 if(state.view==='board'){$('#view').innerHTML='<h1>真实工作，有迹可循</h1><p>沿用现有会话关联，按来源结构分组；不把 Agent 自述完成当成已验收。</p><div class="columns">'+['单来源任务','多来源任务','含执行分支'].map((label,i)=>`<section><h2>${label}</h2>${taskRows().filter(t=>(t.branch_count?2:t.source_count>1?1:0)===i).map(t=>`<button class="card" data-task="${esc(t.id)}">${title(t)}<small>工作状态尚未核实</small></button>`).join('')||'<p>暂无任务</p>'}</section>`).join('')+'</div>';return}
 if(!d){$('#view').innerHTML='<div class="empty">请从左侧选择一个任务。</div>';return}
 const heading='<h1>'+esc(d.title)+'</h1>';
 if(state.view==='story'){
  $('#view').innerHTML=heading+'<p>来源画布 · '+d.source_count+' 份记录 · 尚未生成 AI 故事</p><div class="hint">这里显示真实来源结构，不虚构起因、转折和成果。每份副本与子 Agent 分支均可单独查看。</div><div class="stage">'+d.sources.map(s=>`<button class="card" data-source="${esc(s.id)}"><span class="badge">${esc(relation(s.relation))}</span><h2>${esc(agent(s.agent))}</h2><strong>${esc(s.title)}</strong><small>${s.message_count} 条索引消息 · ${esc(s.id.slice(-8))}</small><small>索引更新 ${esc(s.indexed_at)}</small></button>`).join('')+'</div>';
 }else if(state.view==='map'){
  $('#view').innerHTML=heading+'<p>任务 → 可追溯判断 → 原始依据</p>'+(!d.memory_index_available?'<div class="empty">记忆索引尚不可用；这不代表没有记忆。</div>':!d.memories.length?'<div class="empty">尚无直接引用本任务的记忆。不自动填充示例或推测。</div>':d.memories.map(m=>`<section class="card memory ${esc(m.status)}"><span class="badge">${esc(status(m.status))}</span><h2>${esc(m.subject)}</h2><strong>${esc(m.statement)}</strong><small>成立于 ${esc(m.as_of)} · 私人本地</small><p class="ref">${esc(m.source_ref)}</p><button data-ref="${esc(m.source_ref)}">查看原话依据 →</button></section>`).join(''));
 }else{
  $('#view').innerHTML=heading+'<p>保留每份来源的完整索引顺序，不按相同短句删除多次发生。</p>'+d.sources.map(s=>`<button class="card" data-source="${esc(s.id)}"><strong>${esc(agent(s.agent))} · ${s.message_count} 条索引消息</strong><small>${esc(s.title)} · ${esc(s.relation)}</small></button>`).join('');
 }
}
async function loadProject(){const generation=++state.request;state.messageRequest++;state.project=$('#project').value;state.task='';state.detail=null;state.tasks=[];$('#evidence').hidden=true;render();if(!state.project)return;try{const data=await api('tasks',{project:state.project});if(generation!==state.request)return;state.tasks=data.tasks;render();$('#notice').textContent='真实索引 · 按现有关系归组 · 云端未连接';}catch(e){if(generation===state.request)fail(e)}}
async function selectTask(id){const generation=++state.request;state.messageRequest++;state.task=id;state.detail=null;$('#evidence').hidden=true;render();try{const d=await api('task',{project:state.project,task:id});if(generation!==state.request)return;state.detail=d;if(state.view==='board')state.view='story';render();}catch(e){if(generation===state.request)fail(e)}}
async function evidence(source,offset=0,seq=null){
 const generation=++state.messageRequest;state.source=source;state.seq=seq;
 $('#evidence').hidden=false;
 $('#source').innerHTML=state.detail.sources.map(s=>`<option value="${esc(s.id)}">${esc(agent(s.agent))} · ${esc(relation(s.relation))} · ${esc(s.id.slice(-8))}</option>`).join('');
 $('#source').value=source;$('#messages').textContent='正在读取来源…';$('#page').textContent='';
 $('#prev').disabled=$('#next').disabled=true;
 try{
  const args={project:state.project,task:state.task,source,offset};if(seq)args.sequence=seq;
  const data=await api('messages',args);if(generation!==state.messageRequest)return;
  state.offset=data.offset;state.next=data.next_offset;
  $('#version').textContent='当前索引 · 规则脱敏展示 · 版本 '+data.version;
  $('#messages').innerHTML=data.messages.map(m=>`<article class="${m.sequence===seq?'target':''}"><small>${esc(m.role)} · #${m.sequence} · ${esc(m.occurred_at)}</small><p>${esc(m.text)}</p>${m.truncated?'<small>长消息仅预览前 20,000 字符；未删改原文。</small>':''}<div class="ref">${esc(m.source_ref)}</div></article>`).join('')||'<p>此来源没有索引消息。</p>';
  $('#page').textContent=`${data.total?data.offset+1:0}–${data.offset+data.messages.length} / ${data.total}`;
  $('#prev').disabled=data.offset===0;$('#next').disabled=data.next_offset===null;$('#messages').scrollTop=0;
 }catch(e){if(generation===state.messageRequest){$('#messages').textContent='读取失败，请重新选择来源。';fail(e)}}
}
document.addEventListener('click',e=>{const b=e.target.closest('button');if(!b)return;if(b.dataset.view){state.view=b.dataset.view;render()}if(b.dataset.task)selectTask(b.dataset.task);if(b.dataset.source)evidence(b.dataset.source);if(b.dataset.ref){const match=/^(codex|claude-code):\/\/(.+)#message-(\d+)$/.exec(b.dataset.ref);const s=match&&state.detail.sources.find(s=>s.agent===match[1]&&s.id.replace(/^claude:/,'')===match[2]);if(s)evidence(s.id,0,+match[3]);else $('#notice').textContent='此引用无法在当前任务中定位；没有跨项目查找。'}});
$('#project').onchange=loadProject;$('#search').oninput=render;$('#source').onchange=()=>evidence($('#source').value);$('#close').onclick=()=>{$('#evidence').hidden=true;state.messageRequest++};$('#prev').onclick=()=>evidence(state.source,Math.max(0,state.offset-30));$('#next').onclick=()=>{if(state.next!==null)evidence(state.source,state.next)};
document.addEventListener('keydown',e=>{if(e.key==='Escape')$('#close').click()});
async function init(){
 const generation=++state.request,old=state.project,task=state.task;
 state.messageRequest++;$('#refresh').disabled=true;
 try{
  const data=await api('projects');if(generation!==state.request)return;
  $('#project').innerHTML=data.projects.map(p=>`<option value="${esc(p.id)}">${esc(p.name)} · ${p.identity_kind==='repository'?'仓库':'目录'} · ${p.source_count}</option>`).join('');
  if(data.projects.some(p=>p.id===old))$('#project').value=old;
  const project=$('#project').value;
  await loadProject();
  if(generation+1!==state.request)return;
  if(project===old&&state.tasks.some(t=>t.id===task))await selectTask(task);
  if(!data.projects.length)$('#notice').textContent='还没有已归档项目。请先在 QMemory 桌面端同步。';
 }catch(e){if(generation===state.request)fail(e)}finally{$('#refresh').disabled=false}
}
$('#refresh').onclick=init;init();
